src/textual/widgets/_data_table.py

```diff
@@ -26,40 +29,105 @@
 from ..scroll_view import ScrollView
 from ..strip import Strip
 
+CellCacheKey: TypeAlias = "tuple[RowKey, ColumnKey, Style, bool, bool, int]"
+LineCacheKey: TypeAlias = (
+    "tuple[int, int, int, int, Coordinate, Coordinate, Style, CursorType, bool, int]"
+)
+RowCacheKey: TypeAlias = (
+    "tuple[RowKey, int, Style, Coordinate, Coordinate, CursorType, bool, bool, int]"
+)
 CursorType = Literal["cell", "row", "column", "none"]
-CELL: CursorType = "cell"
 CellType = TypeVar("CellType")
 
 
 class CellDoesNotExist(Exception):
+    """Raised when the user supplies coordinates or cell keys which
+    do not exist in the DataTable."""
+
+    pass
+
+
+@functools.total_ordering
+class StringKey:
+    """An object used as a key in a mapping. It can optionally wrap a string,
+    and lookups into a map using the object behave the same as lookups using
+    the string itself."""
+
+    value: str | None
+
+    def __init__(self, value: str | None = None):
+        self.value = value
+
+    def __hash__(self):
+        # If a string is supplied, we use the hash of the string. If no string was
+        # supplied, we use the default hash to ensure uniqueness amongst instances.
+        return hash(self.value) if self.value is not None else id(self)
+
+    def __eq__(self, other: object) -> bool:
+        # Strings will match Keys containing the same string value.
+        # Otherwise, you'll need to supply the exact same key object.
+        try:
```

Comment:
Comparing the hashes can lead to false positives. It might also make objects compare equal when they are different types, ie. `hash("") == hash(0)`

I think we will need to check it is a valid type to compare it to, and return `NotImplement` if it isn't.